src/webgl/material.js

```diff
@@ -718,171 +725,422 @@ p5.prototype.createFilterShader = function (fragSrc) {
+ *   vec3 lightDir = normalize(uLightDir);
+ *   float brightness = dot(lightDir, normalize(vPosition));
+ *   brightness = clamp(brightness, 0.4, 1.0);
+ *   vec3 color = vec3(0.3, 0.5, 1.0);
+ *   color = color * brightness * 3.0;
+ *   gl_FragColor = vec4(color, 1.0);
  * }
+ * `;
  *
  * function setup() {
  *   createCanvas(100, 100, WEBGL);
+ *   fillShader = createShader(vertSrc, fragSrc);
+ *   noStroke();
+ *   describe('A rotating torus with simulated directional lighting.');
+ * }
  *
- *   // Use the p5.Shader object.
- *   shader(mandelbrot);
+ * function draw() {
+ *   background(20, 20, 40);
+ *   let lightDir = [0.5, 0.5, -1.0];
+ *   fillShader.setUniform('uLightDir', lightDir);
+ *   shader(fillShader);
+ *   rotateY(frameCount * 0.02);
+ *   rotateX(frameCount * 0.02);
+ *   //lights();
+ *   torus(25, 10, 30, 30);
+ * }
+ * </code>
+ * </div>
  *
- *   // Set the shader uniform p to an array.
- *   mandelbrot.setUniform('p', [-0.74364388703, 0.13182590421]);
+ * @example
+ * <div modernizr='webgl'>
+ * <code>
+ * let fillShader;
  *
- *   describe('A fractal image zooms in and out of focus.');
+ * let vertSrc = `
+ * precision highp float;
+ * attribute vec3 aPosition;
+ * uniform mat4 uProjectionMatrix;
+ * uniform mat4 uModelViewMatrix;
+ * varying vec3 vPosition;
+ * void main() {
+ *   vPosition = aPosition;
+ *   gl_Position = uProjectionMatrix * uModelViewMatrix * vec4(aPosition, 1.0);
  * }
+ * `;
  *
- * function draw() {
- *   // Set the shader uniform r to a value that oscillates between 0 and 2.
- *   mandelbrot.setUniform('r', sin(frameCount * 0.01) + 1);
+ * let fragSrc = `
+ * precision highp float;
+ * uniform vec3 uLightPos;
+ * uniform vec3 uFillColor;
+ * varying vec3 vPosition;
+ * void main() {
+ *   float brightness = dot(normalize(uLightPos), normalize(vPosition));
+ *   brightness = clamp(brightness, 0.0, 1.0);
+ *   vec3 color = uFillColor * brightness;
+ *   gl_FragColor = vec4(color, 1.0);
+ * }
+ * `;
  *
- *   // Add a quad as a display surface for the shader.
- *   quad(-1, -1, 1, -1, 1, 1, -1, 1);
+ * function setup() {
+ *   createCanvas(100, 100, WEBGL);
+ *   fillShader = createShader(vertSrc, fragSrc);
+ *   shader(fillShader);
+ *   noStroke();
+ *   describe('A square affected by both fill color and lighting, with lights controlled by mouse.');
+ * }
+ *
+ * function draw() {
+ *   let lightPos = [(mouseX - width / 2) / width, (mouseY - height / 2) / height, 1.0];
+ *   fillShader.setUniform('uLightPos', lightPos);
+ *   let fillColor = [map(mouseX, 0, width, 0, 1), map(mouseY, 0, height, 0, 1), 0.5];
+ *   fillShader.setUniform('uFillColor', fillColor);
+ *   plane(100, 100);
  * }
  * </code>
  * </div>
+ */
+p5.prototype.shader = function (s) {
+  this._assert3d('shader');
+  p5._validateParameters('shader', arguments);
+
+  s.ensureCompiledOnContext(this);
+
+  // Always set the shader as a fill shader
+  this._renderer.userFillShader = s;
+  this._renderer._useNormalMaterial = false;
+
```

Comment:
I think you can still keep this line 
```js
s.setDefaultUniforms();
```
since it's used by `shaderHooks`. for every entry of `this.hooks.uniforms` it sets the default values. So, we could keep it.